apps/service_providers/llm_service/adapters.py

```diff
@@ -119,17 +93,176 @@ def save_message_to_history(self, message: str, type_: ChatMessageType, experime
+    def from_pipeline(experiment: Experiment, session: ExperimentSession) -> Self:
+        """TODO"""
+
+    def get_chat_model(self):
+        return self.get_llm_service().get_chat_model(self.provider_model_name, self.temperature)
+
+    def get_template_context(self, variables: list[str]):
+        return self.template_context.get_context(variables)
+
+    def get_prompt(self):
+        return self.prompt_text
+
+    def get_chat_history(self, input_messages: list):
+        return compress_chat_history(
+            self.session.chat,
+            llm=self.get_chat_model(),
+            max_token_limit=self.max_token_limit,
+            input_messages=input_messages,
+        )
+
     def check_cancellation(self):
-        self.chat.refresh_from_db(fields=["metadata"])
+        self.session.chat.refresh_from_db(fields=["metadata"])
         # temporary mechanism to cancel the chat
         # TODO: change this to something specific to the current chat message
-        return self.chat.metadata.get("cancelled", False)
+        return self.session.chat.metadata.get("cancelled", False)
+
+
+class AssistantAdapter(BaseAdapter):
+    def __init__(
+        self,
+        session: ExperimentSession,
+        assistant: OpenAiAssistant,
+        citations_enabled: bool,
+        experiment_version_number: int,
+        experiment_is_a_version: bool,
+        input_formatter: str | None = None,
+        trace_service=None,
+        save_message_metadata_only: bool = False,
+    ):
+        self.session = session
+        self.team = session.team
+        self.assistant = assistant
+        self.llm_service = assistant.llm_provider.get_llm_service()
+        self.provider_model_name = assistant.llm_provider_model.name
+        self.citations_enabled = citations_enabled
+        self.experiment_version_number = experiment_version_number
+        self.experiment_is_a_version = experiment_is_a_version
+        self.tools = get_assistant_tools(assistant, experiment_session=session)
+        self.input_formatter = input_formatter
+        self.trace_service = trace_service
+        self.save_message_metadata_only = save_message_metadata_only
+        self.input_message_metadata = {}
+        self.output_message_metadata = {}
+        self.template_context = PromptTemplateContext(session, source_material_id=None)
+
+    @staticmethod
+    def from_experiment(experiment: Experiment, session: ExperimentSession, trace_service=None) -> Self:
+        return AssistantAdapter(
+            session=session,
+            assistant=experiment.assistant,
+            citations_enabled=experiment.citations_enabled,
+            experiment_version_number=experiment.version_number,
+            experiment_is_a_version=experiment.is_a_version,
+            input_formatter=experiment.input_formatter,
+            trace_service=trace_service,
+        )
+
+    @staticmethod
+    def from_pipeline(session: ExperimentSession, node: "AssistantNode") -> Self:
```

Comment:
Should this be called `from_pipeline_node`?

Alternatively, we could rename them `for_experiment` and `for_pipeline`

Reply:
I like `for_experiment`. Let's go with that

Reply:
https://github.com/dimagi/open-chat-studio/pull/918/commits/930a80a13eb58f0919231e398cf1a51d0f39fcd0